### PR TITLE
Fixed #4396, #4883 : Add CI checks to ensure Kotlin files , Layout xml files are only referencing colors from component_colors.xml

### LINKED ISSUE #4396: Add CI checks to ensure layout xml files are only referencing colors from component_colors.xml
According to newer color conventions, all layout files should only access colors from `component_colors.xml`. 
Add a CI check ensuring the same.

Reference: #4203, #4205

### LINKED ISSUE #4883: Add CI checks to ensure Kotlin files are only referencing colors from component_colors.xml
According to newer color conventions, all Kotlin files (Activities, Activity Presenters, Fragments, Fragment Presenters, Views, and View Presenters) should only access colors from `component_colors.xml`.
Add a CI check ensuring the same.

### PR DESCRIPTION
**Pull request by MohitGupta121**
## Explanation

Fixed #4396 : Add CI checks to ensure Layout xml files are only referencing colors from component_colors.xml 

Fixed #4883 : Add CI checks to ensure Kotlin Files (Activities, Activity Presenters, Fragments, Fragment Presenters, Views, and View Presenters) are only referencing colors from component_colors.xml 

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Review by MohitGupta121 — COMMENTED** [review]
@BenHenning PTAL, the check is working fine for all xml files as you see in the CI check [failure](https://github.com/oppia/oppia-android/actions/runs/4263499965/jobs/7420315948#step:7:102).
And also Kotlin files checks are [failing](https://github.com/oppia/oppia-android/actions/runs/4264049569/jobs/7421592111#step:7:102).

File: scripts/assets/file_content_validation_checks.textproto [hunk 1]
```diff
@@ -339,6 +339,31 @@ file_content_checks {
   prohibited_content_regex: "name=.+(?<!_color).>@|name=.*[A-Z].*>@"
   failure_message: "All color declarations in component_color.xml and color_palette.xml should end with _color suffix following snake_case naming convention."
 }
+file_content_checks {
+  file_path_regex: "app/src/main/res/layout.*?/.+?\\.xml"
+  prohibited_content_regex: "@color/(?!component_color_).+"
+  failure_message: "Only colors from component_colors.xml may be used in layouts."
+}
+file_content_checks {
+  file_path_regex: "app/src/main/java/org/oppia/android/app.*?/.+ActivityPresenter\\.kt"
```
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto]: You may be able to combine this & the others into a single check for Kotlin files. We also need to add View and ViewPresenter.

**Review by BenHenning — COMMENTED** [review]
This looks reasonable @MohitGupta121. Is there anything you want me to look at besides the patterns?

File: scripts/assets/file_content_validation_checks.textproto (hunk 1 above)
Comment (MohitGupta121) [reply scripts/assets/file_content_validation_checks.textproto]: Yes, I also thinking to combine them but to getting the pattern of path, how to write expression that enables for all of these?

Yes we need View ans ViewPresenter, But for View file "View" Present as prefix and ViewPresenter as suffix. So we need two checks for that?

**Comment by MohitGupta121** [comment]
>  Is there anything you want me to look at besides the patterns?

I don't think anything else expect the Path_Pattern. And nothing from the issue side also as I see.

@BenHenning What do you think ?

File: scripts/assets/file_content_validation_checks.textproto (hunk 1 above)
Comment (MohitGupta121) [reply scripts/assets/file_content_validation_checks.textproto]: Done.

**Comment by MohitGupta121** [comment]
@BenHenning PTAL, Combine kotlin files checks into 1 check
And not find any file whose suffix is ViewPresenter, So I keep it as it.
But we have files whose prefix is View but unable to get expression pattern for that. So can you please help in that.

File: scripts/assets/file_content_validation_checks.textproto (hunk 1 above)
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto]: Sorry I'm not sure that I understand the question for views--isn't it the case that all of our custom views end with 'View'? We don't currently have view presenters today as far as I know, but that's technically a pattern that's possible (and something we'll move toward once we introduce some of our planned code generation for UI code).

**Review by BenHenning — COMMENTED** [review]
Thanks @MohitGupta121. Do you have examples for views that use 'View' as a prefix that would need updating?

Also, please update the PR description to have the fixed issues as separate lines (otherwise GitHub doesn't hook them up to automation correctly).

**Comment by MohitGupta121** [comment]
@BenHenning Yes got the views whose prefix is "View" https://github.com/oppia/oppia-android/tree/develop/app/src/main/java/org/oppia/android/app/customview

**Comment by MohitGupta121** [comment]
@BenHenning updated Tests for CustomViews also. Here is examples for all tests that are failing :

- Layout xml files :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:132

-  Kotlin File : 

1. Activity .kt :  
2. Fragments .kt : 
3. ActivityPresenter .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:131
4. FragmentPresenter .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:107
5. View .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:106

At present view not have ViewPresenter so not have any example as such but added Test in regex.

**Comment by MohitGupta121** [comment]
@BenHenning PTAL, 
Updated Tests for CustomViews also. Here is examples for all tests that are failing :

- Layout xml files :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:132

-  Kotlin Files : 

1. Activity .kt :  https://github.com/oppia/oppia-android/actions/runs/4289352575/jobs/7472301291#step:7:128
2. Fragments .kt :  https://github.com/oppia/oppia-android/actions/runs/4289352575/jobs/7472301291#step:7:129
3. ActivityPresenter .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:131
4. FragmentPresenter .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:107
5. View .kt :  https://github.com/oppia/oppia-android/actions/runs/4289267205/jobs/7472132371#step:7:106

At present view not have ViewPresenter so not have any example as such but added Test in regex.

**Comment by MohitGupta121** [comment]
@BenHenning we need to fix the Regex checks that are failing before this PR merged? Or first you give your review?

**Review by BenHenning — COMMENTED** [review]
Thanks @MohitGupta121! The patterns look good--can you also add tests to verify the patterns (similar to how we do for all other file content patterns)?

**Comment by BenHenning** [comment]
> @BenHenning we need to fix the Regex checks that are failing before this PR merged? Or first you give your review?

@MohitGupta121 what failures are there (the logs have since expired)? If there are just a few, they could be fixed right in this PR. If there are a bunch, it may be better to have an earlier PR.

**Comment by MohitGupta121** [comment]
> Thanks @MohitGupta121! The patterns look good--can you also add tests to verify the patterns (similar to how we do for all other file content patterns)?

@BenHenning Here in this `RegexPatternValidationCheckTest.kt` we need to add Patterns ? Here also do we need to add 2 separate Tests for Layouts and Xml's ? 

What patterns we need to check such as any example?

**Comment by MohitGupta121** [comment]
> @MohitGupta121 what failures are there (the logs have since expired)? If there are just a few, they could be fixed right in this PR. If there are a bunch, it may be better to have an earlier PR.

@BenHenning From [here](https://github.com/oppia/oppia-android/actions/runs/4390743622/jobs/7689426030#step:7:102) till the bottom the checks are failing? So what you suggests after looking into it, we solve in this PR or new followup PR?

**Comment by MohitGupta121** [comment]
And @BenHenning How we can tests `RegexPatternValidationCheckTest` locally? Is there any command similar to Regex static checks?

**Comment by BenHenning** [comment]
> > Thanks @MohitGupta121! The patterns look good--can you also add tests to verify the patterns (similar to how we do for all other file content patterns)?
> 
> @BenHenning Here in this `RegexPatternValidationCheckTest.kt` we need to add Patterns ? Here also do we need to add 2 separate Tests for Layouts and Xml's ?
> 
> What patterns we need to check such as any example?

As discussed in meeting, we should have at least 2 tests (one for each pattern), and those tests should cover all the key cases tested by the patterns (including verifying that files which should have the pattern off do actually have it off). Examples can be seen here: https://github.com/oppia/oppia-android/blob/develop/scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt.

> > @MohitGupta121 what failures are there (the logs have since expired)? If there are just a few, they could be fixed right in this PR. If there are a bunch, it may be better to have an earlier PR.
> 
> @BenHenning From [here](https://github.com/oppia/oppia-android/actions/runs/4390743622/jobs/7689426030#step:7:102) till the bottom the checks are failing? So what you suggests after looking into it, we solve in this PR or new followup PR?

As discussed in meeting, it's up to you. You could either merge this with those files as exemptions and fix them in a follow-up, or fix them before merging this PR. The former is probably a bit cleaner since it means fewer PRs in review, and it ensures all the other files in the codebase that are passing stay passing.

**Comment by MohitGupta121** [comment]
@BenHenning  in `RegexPatternValidationCheckTest.kt`  ca we use File Path like this `app/src/main/res/layout.*?/.+?\\.xml` instead of exact file path as we have so many layouts files, So we not mention all layouts file and same for Kotlin `.kt ` files in Tests ?

**Review by MohitGupta121 — COMMENTED** [review]
@BenHenning PTAL, I updated the Regex Validation Tests along with exceptions files.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt [hunk 2]
```diff
@@ -2153,6 +2158,98 @@ class RegexPatternValidationCheckTest {
       )
   }
 
+  fun testFileContent_xmlLayouts_includesNonColorComponentReferences_fileContentIsNotCorrect() {
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:2162]: This & the test below are missing ``@Test`` annotations. Just to check, did you verify that these tests fail when the new checks are temporarily removed?

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt [hunk 3]
```diff
@@ -2153,6 +2158,98 @@ class RegexPatternValidationCheckTest {
...
+
+  fun testFileContent_kotlinFiles_includesNonColorComponentReferences_fileContentIsNotCorrect() {
+    val prohibitedContent =
+      """
+        R.color.component_color_shared_activity_status_bar_color
+        R.color.color_def_avatar_background_1
+        R.color.color_palette_primary_color
+      """.trimIndent()
+    tempFolder.newFolder("testfiles", "app", "src", "main", "java", "org", "oppia", "android")
+    val stringFilePath = "app/src/main/java/org/oppia/android/app.?/." +
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Does this actually work? I didn't think newFile() works with patterns.

I think for this case we should be creating an example file for each type that should be prohibited (which means that multiple filepaths will be present in the failure text).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MohitGupta121! Just had a few suggestions--PTAL.

> @BenHenning in `RegexPatternValidationCheckTest.kt` ca we use File Path like this `app/src/main/res/layout.*?/.+?\\.xml` instead of exact file path as we have so many layouts files, So we not mention all layouts file and same for Kotlin `.kt ` files in Tests ?

I see that you added the explicit files--this is the preferred way to do it since we want to try and prevent the prohibited pattern from continuing.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt (hunk 2 above)
Comment (MohitGupta121) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:2162]: Okay I'll add @Test, Yes Ben, that are failing both in my local as well as in GitHub workflow.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt (hunk 3 above)
Comment (MohitGupta121) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: How I can create example file? Is that manually like we add new file or is there any script through test itself create one?

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt [hunk 4]
```diff
@@ -2153,6 +2158,148 @@ class RegexPatternValidationCheckTest {
...
+
+  @Test
+  fun testFileContent_kotlinFiles_includesNonColorComponentReferences_fileContentIsNotCorrect() {
+    val prohibitedContent =
+      """
+        R.color.component_color_shared_activity_status_bar_color
+        R.color.color_def_avatar_background_1
+        R.color.color_palette_primary_color
+      """.trimIndent()
+    tempFolder.newFolder(
```
Comment (MohitGupta121) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:2224]: @BenHenning I used here this format as the path crossing the lint limit "Exceeded max line length (100)".
That's why I arrange in this order, Is it good or there is any other way to fix the lint?

**Review by MohitGupta121 — COMMENTED** [review]
@BenHenning PTAL, Thanks.
I updated all the Regex Tests and now working for all kotlin as well as xml.
Please tell me if we need to change or update anything else.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt (hunk 4 above)
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:2224]: Nope, this looks like the correct way to format it.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt (hunk 3 above)
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Latest changes have addressed this.

**Review by BenHenning — APPROVED** [review]
Thanks @MohitGupta121! This LGTM!

**Pull request merged**